fast_llm/layers/language_model/multi_token_prediction_head.py

```diff
@@ -0,0 +1,132 @@
+import typing
+
+import torch
+from torch._C._distributed_c10d import ReduceOp  # noqa
+
+from fast_llm.engine.config_utils.tensor_space import DefaultDimNames, TensorSpace
+from fast_llm.engine.distributed.config import DistributedDimNames
+from fast_llm.layers.common.auxiliary_loss import AuxiliaryLoss
+from fast_llm.layers.language_model.config import LanguageModelBaseConfig, LanguageModelKwargs, LanguageModelLossNames
+from fast_llm.layers.language_model.embedding import WORD_EMBEDDINGS_WEIGHT
+from fast_llm.layers.language_model.head import OUTPUT_WEIGHTS, LanguageModelHead
+from fast_llm.layers.transformer.config import TransformerKwargs
+from fast_llm.layers.transformer.transformer import TransformerLayer
+from fast_llm.tensor import TensorMeta
+
+
+class MultiTokenPredictionTransformerLayer(TransformerLayer):
+    """
+    A transformer layer for multi-token prediction.
+    Takes a shared_hidden as input.
+    Returns a stack of shared_hidden and transformer_output.
+    """
+
+    # TODO MTP: what layer-index for transformer layer? All the transformer layers in the MTP-LM-heads
+    # will have the same name. Is it an issue?
+    def forward(
+        self,
+        input_: torch.Tensor,
+        kwargs: dict[str, typing.Any],
+        losses: dict[str, typing.Any] | None = None,
+        metrics: dict[str, typing.Any] | None = None,
+    ) -> torch.Tensor:
+        # TODO MTP: fix dims?
+        if isinstance(input_, TensorMeta):
+            return self._get_meta(input_, "output", kwargs)
+        output = super().forward(input_, kwargs, losses, metrics)
+        return torch.stack((input_, output), dim=0)
+
+
+class MultiTokenPredictionLanguageModelHead(LanguageModelHead):
```

Comment:
Is this really worth a separate class? Seems like we could just add a `multi_token_prediction_index`/`_prediction_distance` option to the base class, it would only lead to tiny changes.

Reply:
Hmm, we could yes, but there are quite a bit of changes, especially in `_foward_backward`, with the shifting/truncating of labels and inputs, and the potential handling of sequence-parallel and cross-entropy-splits (not supported now, but this would need different code).
Do you still think we should put this in the base class `LanguageModelHead` ?

Reply:
Oh I guess the logic of the base class would just be a special-case where `multi_token_prediction_index=0`
Then indeed I could just add this code to the base class

Reply:
I could also remove the `MultiTokenPredictionTransformerLayer` class by adding a `stacked_output` parameter to the `TransformerLayer` class. wdyt @jlamypoirier @tscholak ?

Reply:
I removed both `MultiTokenPredictionTransformerLayer` and `MultiTokenPredictionLanguageModelHead` and moved the code to the base classes. Things should be a bit simpler now.